fix(mikro-orm): re-enable tests

boredland opened the pull request:
swc/jest and decorators seems to be a hard nut to crack. But probably having reflection based metadata on our schema (and thus putting assumptions on the users setup) has not been a good idea to begin with. So I decided to follow the approach in an [upstream PR](https://github.com/mikro-orm/mikro-orm/pull/3200/files) and inserted reflection-less meta-data everywhere.

## ☕️ Reasoning

Aims to re-enable the tests, as I'd like to make some changes and need working regression tests.

closes https://github.com/nextauthjs/next-auth/issues/5286 as a side-effect.

## 🧢 Checklist

- [ ] Documentation
- [x] Tests
- [x] Ready to be merged

boredland committed:
fix(mikro-orm): re-enable tests by setting up ts-jest

While I have a working setup for tests with swc using 'next/jest',
I struggle getting a working setup here. So I installed ts-jest and some
missing peer dependencies.

boredland committed:
refactor(mikro-orm): explicit type annotations

boredland committed:
refactor(mikro-orm): use named references for mapped entities

boredland committed:
chore(mikro-orm): simplify testing setup

boredland committed:
fix(mikro-orm): make id_token text and providertype string

BREAKING CHANGE:

apparently oauth providers were not working previously,
as two important fields had the wrong types.

this change will require users to run an appropriate migration.

closes https://github.com/nextauthjs/next-auth/issues/5286

boredland committed:
chore(mikro-orm): update lockfile

balazsorban44 commented on packages/adapter-mikro-orm/package.json:
```diff
@@ -32,22 +32,21 @@
   ],
   "peerDependencies": {
-    "@mikro-orm/core": "^5.0.2",
+    "@mikro-orm/core": "^5.4.1",
```
I don't think we should bump peer dependencies, because it will error on some package managers. Is this really a necessary change? If so, this will be a major bump IMO.

boredland replied:
I don't think its strictly necessary, but will run a test.

boredland committed:
fix(mikro-orm): loosen peerdependency to core

boredland commented:
I think I just had a nice idea on how to test for breaking schema changes - will add this here.

boredland committed:
test(mikro-orm): snaphot of schema and create statement

boredland commented:
@balazsorban44 as we previously had a broken schema that was unusable for sso providers as I see it, we should make this a breaking change - users will need to run a migration that updates fields to their new types.

boredland commented:
I think all issues are resolved and CI is happy. Would love to get this merged soon @balazsorban44 :-)

ThangHuuVu reviewed (approved):
I skimmed the PR, it looks pretty good to me. Before we merge it, I'd like t ask for @ndom91's review

boredland commented:
Is there anything I can do to speed this up?

boredland-ioki commented:
@ndom91 and @balazsorban44 should I rebase this? don't want it to go stale again...

boredland committed:
fix(mikro-orm): ensure collection type

BREAKING CHANGE

in mikro-orm prior to 4.2 collections emit a type `Collection<Session,
unknown>` this makes them essentially incompatible (read: unextendable)
by the newest mikro-orm versions. In newer versions this always should
be `Collection<Session, object>`, hence enforcing this for older mikro-
orm versions.

boredland committed:
chore(mikro-orm): upgrade uuid and jest

boredland commented:
added another fix on top and updated some versions. perhaps you could have another look @ThangHuuVu ?

ndom91 commented:
Yeah this also looks good to me! Just approved another round of CI to double check the tests pass :+1:

ThangHuuVu commented:
Cool, thank you @boredland @ndom91, merging it now 🙌
